lib/Conversion/LinalgToArith/LinalgToArith.cpp

```diff
@@ -0,0 +1,385 @@
+                                leftOperandSecretness.getSecretness());
+
+    auto *rightOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[1]);
+    Secretness rightOperandSecretness =
+        rightOperandLookup ? rightOperandLookup->getValue() : Secretness();
+    bool rightOperandIsSecret = (rightOperandSecretness.isInitialized() &&
+                                 rightOperandSecretness.getSecretness());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Left operand is secret: " << leftOperandIsSecret << "\n"
+                   << "Right operand is secret: " << rightOperandIsSecret
+                   << "\n";
+    });
+
+    // Error out if both are secret or both are public
+    if ((leftOperandIsSecret && rightOperandIsSecret) ||
+        (!leftOperandIsSecret && !rightOperandIsSecret)) {
+      return failure();
+    }
+    auto inputs = op.getInputs();
+    auto outputs = op.getOutputs();
+
+    Value secretValues;
+    Value publicMatrix;
+    if (leftOperandIsSecret) {
+      secretValues = inputs[0];
+      publicMatrix = inputs[1];
+    } else {  // the right operand is secret
+      secretValues = inputs[1];
+      publicMatrix = inputs[0];
+    }
+    auto matrixTensorType = cast<RankedTensorType>(publicMatrix.getType());
+    auto bias = outputs[0];
+
+    auto dimensions = matrixTensorType.getShape();
+    int64_t dim0 = dimensions[0];  // This is the number of rows in the matrix.
+    int64_t dim1 = dimensions[1];  // This is the number of columns in the
+    // matrix.
+
+    SmallVector<int64_t> transposedDimensions({dim1, dim0});
+
+    // If one of these dimensions is not a power of two, then we can't do the
+    // Halevi-Shoup or Squat Packing Matrix Multiplication conversion.
+    if (!isPowerOfTwo(dim0) || !isPowerOfTwo(dim1)) {
+      return failure();
+    }
+
+    // If the matrix is not a square matrix, then we are doing squat packing.
+    // If so, then the vector dimension is also the longer dimension.
+    if ((leftOperandIsSecret && dim0 < dim1) ||
+        (rightOperandIsSecret && dim0 > dim1)) {
+      return failure();
+    }
+
+    // Diagonalize the matrix first.
+    auto constantValues =
+        dyn_cast<arith::ConstantOp>(publicMatrix.getDefiningOp());
+    if (!constantValues) {
+      return failure();
+    }
+
+    DenseElementsAttr denseAttr =
+        dyn_cast<DenseElementsAttr>(constantValues.getValueAttr());
+
+    // If the constant values doesn't have a dense attribute, then we can't
+    // diagonalize the matrix.
+    if (!denseAttr) {
+      return failure();
+    }
+
+    auto type = denseAttr.getElementType();
+    bool intType = llvm::isa<IntegerType>(type);
+    bool floatType = llvm::isa<FloatType>(type);
+    if (!intType && !floatType) {
+      return failure();
+    }
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    Value newMatrix;
+
+    // Modify the matrix to be a diagonal matrix. We'll simply create a copy of
```

Comment:
Maybe factor this code into a helper like 
```
template <typename T>
FailureOr<Value> diagonalizeMatrix(DenseElementsAttr constantAttr, bool transpose, ImplicitLocOpBuilder builder) {
      auto type = constantAttr.getElementType();

     auto [dim0, dim1] = constantAttr.getShape();
     SmallVector<int64_t> transposedDimensions({dim1, dim0});

      SmallVector<T> diagonalElements;
      diagonalElements.reserve(dim0 * dim1);
      for (int i = 0; i < transposedDimensions[0]; ++i) {
        for (int j = 0; j < transposedDimensions[1]; ++j) {
          int index =
              calculateIndexHelper(transpose, dim0, dim1, i, j);
          auto value = denseAttr.getValues<T>()[index];
          diagonalElements.push_back(value);
        }
      }
      auto diagonalizedType = RankedTensorType::get(transposedDimensions, type);
      auto diagonalizedDenseElementsAttr =
          DenseElementsAttr::get(diagonalizedType, diagonalElements);
      newMatrix = b.create<arith::ConstantOp>(diagonalizedDenseElementsAttr);
}
```

and call with:
```
    if (type.isIntType()) {
      newMatrix = diagonalizeMatrix<APInt>(denseAttr, leftOperandIsSecret, b);
    } else {
      newMatrix = diagonalizeMatrix<APFloat>(denseAttr, leftOperandIsSecret, b);
```

I think that will resolve the template issue...